spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreNextPageAwareResultsIterable.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2019-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.datastore.core;
+
+import java.util.Iterator;
+
+import com.google.cloud.datastore.Cursor;
+
+/**
+ * An iterable that contains a cursor for the next page and can be used to determine
+ * if the next page exists.
+ *
+ * @author Dmitry Solomakha
+ */
+public class DatastoreNextPageAwareResultsIterable<T> implements Iterable<T> {
```

Comment:
I don't think this is quite the right approach; it seems a bit heavyweight to create a new class to hold a boolean.

Also the only method that returns this is `nextPageAwareQuery`; however, other query methods such as `query`, `queryIterable`, and `queryByExample` also return `DatastoreResultsIterable` but are not able to take advantage of the next-page-aware feature.

I would try to get `DatastoreResultsIterable` to have access to the `hasNextPage` somehow.

I would suggest a couple alternatives:

1. Is it possible to add a method to DatastoreTemplate like `hasNextPage(Cursor c)` or `hasNextPage(DatastoreResultsIterable i)`? if so, the user can manually check for next page through usage of template.

2. If the query is cheap, you might as well pay the cost of `nextPageExists(..)` for each DatastoreResultsIterable; and build the `hasNextPage` parameter in there. no need for premature optimization

3. Add a method called `hasNextPage()` on the `DatastoreResultsIterable` which will run the query of checking if the cursor has a next page. This would require the Iterable obj have reference to the datastore client I suppose.

Reply:
+1
Can we just extend `DatastoreResultsIterable`?